File: backend/src/main/java/com/cibertec/registroempleados/service/ValidationService.java
package com.cibertec.registroempleados.service;

import org.springframework.stereotype.Service;
import java.util.HashMap;
import java.util.Map;

@Service
public class ValidationService {

    // Email validation
    public boolean isValidEmail(String email) {
        if (email == null || email.isEmpty()) return false;
        String emailRegex = "^[^\\s@]+@[^\\s@]+\\.[^\\s@]+$";
        return email.matches(emailRegex);
    }

    // Phone validation (Peru format)
    public boolean isValidPhone(String phone) {
        if (phone == null || phone.isEmpty()) return false;
        String phoneRegex = "^(\\+51|0)?[1-9]\\d{8,9}$";
        return phone.replaceAll("\\s", "").matches(phoneRegex);
    }

    // DNI validation (Peru format)
    public boolean isValidDNI(String dni) {
        if (dni == null || dni.isEmpty()) return false;
        return dni.matches("^\\d{8}$");
    }

    // Not empty validation
    public boolean isNotEmpty(String value) {
        return value != null && !value.trim().isEmpty();
    }

    // Min length validation
    public boolean minLength(String value, int length) {
        return value != null && value.length() >= length;
    }

    // Max length validation
    public boolean maxLength(String value, int length) {
        return value != null && value.length() <= length;
    }

    // Generic validation with multiple rules
    public Map<String, String> validateEmpleado(String nombres, String apellidos, String email, String telefono, String dni) {
        Map<String, String> errors = new HashMap<>();

        if (!isNotEmpty(nombres)) {
            errors.put("nombres", "El nombre es requerido");
        } else if (!minLength(nombres, 2)) {
            errors.put("nombres", "El nombre debe tener al menos 2 caracteres");
        }

        if (!isNotEmpty(apellidos)) {
            errors.put("apellidos", "Los apellidos son requeridos");
        } else if (!minLength(apellidos, 2)) {
            errors.put("apellidos", "Los apellidos deben tener al menos 2 caracteres");
        }

        if (!isNotEmpty(email)) {
            errors.put("email", "El email es requerido");
        } else if (!isValidEmail(email)) {
            errors.put("email", "El email es inválido");
        }

        if (!isNotEmpty(telefono)) {
            errors.put("telefono", "El teléfono es requerido");
        } else if (!isValidPhone(telefono)) {
            errors.put("telefono", "El teléfono es inválido");
        }

        if (!isNotEmpty(dni)) {
            errors.put("dni", "El DNI es requerido");
        } else if (!isValidDNI(dni)) {
            errors.put("dni", "El DNI debe tener 8 dígitos");
        }

        return errors;
    }
}

File: frontend/src/app/services/validation.service.ts
import { Injectable } from '@angular/core';
import { AbstractControl, ValidationErrors } from '@angular/forms';

@Injectable({
  providedIn: 'root'
})
export class ValidationService {
  constructor() {}

  // Email validation
  isValidEmail(email: string): boolean {
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    return emailRegex.test(email);
  }

  // Phone validation (Perú format)
  isValidPhone(phone: string): boolean {
    const phoneRegex = /^(\+51|0)?[1-9]\d{8,9}$/;
    return phoneRegex.test(phone.replace(/\s/g, ''));
  }

  // DNI validation (Peru format)
  isValidDNI(dni: string): boolean {
    const dniRegex = /^\d{8}$/;
    return dniRegex.test(dni);
  }

  // Password strength validation
  isStrongPassword(password: string): boolean {
    // At least 8 characters, 1 uppercase, 1 lowercase, 1 number, 1 special char
    const strongRegex = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/;
    return strongRegex.test(password);
  }

  // Required field validation
  isNotEmpty(value: any): boolean {
    if (typeof value === 'string') {
      return value.trim().length > 0;
    }
    return value != null && value !== '';
  }

  // Min length validation
  minLength(value: string, length: number): boolean {
    return !!value && value.length >= length;
  }

  // Max length validation
  maxLength(value: string, length: number): boolean {
    return !!value && value.length <= length;
  }

  // Number validation
  isNumber(value: any): boolean {
    return !isNaN(value) && value !== '';
  }

  // Date validation
  isValidDate(dateString: string): boolean {
    const date = new Date(dateString);
    return date instanceof Date && !isNaN(date.getTime());
  }

  // URL validation
  isValidUrl(url: string): boolean {
    try {
      new URL(url);
      return true;
    } catch {
      return false;
    }
  }

  // Get validation error message
  getErrorMessage(control: AbstractControl | null, fieldName: string): string {
    if (!control) return '';

    if (control.hasError('required')) {
      return `${fieldName} es requerido`;
    }
    if (control.hasError('minlength')) {
      return `${fieldName} debe tener al menos ${control.getError('minlength').requiredLength} caracteres`;
    }
    if (control.hasError('maxlength')) {
      return `${fieldName} no puede exceder ${control.getError('maxlength').requiredLength} caracteres`;
    }
    if (control.hasError('email')) {
      return `${fieldName} debe ser un email válido`;
    }
    if (control.hasError('pattern')) {
      return `${fieldName} tiene un formato inválido`;
    }

    return '';
  }
}
